Build error with clang-19

### Generated by Generative AI

_No response_

### Operating System:

Ubuntu 24.04

### ROS version or commit hash:

Rolling

### RMW implementation (if applicable):

_No response_

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

_No response_

### 'ros2 doctor --report' output

<details><summary>ros2 doctor --report</summary>

```console
<COPY OUTPUT HERE>
```
</details>


### Steps to reproduce issue

sudo apt install clang-19
CC=/usr/bin/clang-19 CXX=/usr/bin/clang++-19 colcon build --packages-up-to rmw_zenoh_cpp

### Expected behavior

Build succeeds

### Actual behavior

```
--- stderr: rmw_zenoh_cpp                             
/home/yadunund/ros2_rolling/src/thirdparty/rmw_zenoh/rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp:129:42: error: no matching member function for call to 'emplace'
  129 |       adv_pub_opts.sample_miss_detection.emplace({}).heartbeat =
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/optional:928:2: note: candidate template ignored: couldn't infer template argument '_Up'
  928 |         emplace(initializer_list<_Up> __il, _Args&&... __args)
      |         ^
/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/optional:916:2: note: candidate template ignored: substitution failure: deduced incomplete pack <(no value)> for template parameter '_Args'
  913 |       template<typename... _Args>
      |                            ~~~~~
  914 |         _GLIBCXX20_CONSTEXPR
  915 |         enable_if_t<is_constructible_v<_Tp, _Args...>, _Tp&>
  916 |         emplace(_Args&&... __args)
      |         ^
/home/yadunund/ros2_rolling/src/thirdparty/rmw_zenoh/rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp:203:29: error: no matching member function for call to 'emplace'
  203 |       adv_sub_opts.recovery.emplace({}).last_sample_miss_detection =
      |       ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/optional:928:2: note: candidate template ignored: couldn't infer template argument '_Up'
  928 |         emplace(initializer_list<_Up> __il, _Args&&... __args)
      |         ^
/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/optional:916:2: note: candidate template ignored: substitution failure: deduced incomplete pack <(no value)> for template parameter '_Args'
  913 |       template<typename... _Args>
      |                            ~~~~~
  914 |         _GLIBCXX20_CONSTEXPR
  915 |         enable_if_t<is_constructible_v<_Tp, _Args...>, _Tp&>
  916 |         emplace(_Args&&... __args)
      |         ^
1 error generated.
```

### Additional information

llvm clang seems to be stricter than gcc. It's unable to find the default constructor.

--- a/rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp
+++ b/rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp
@@ -115,6 +115,7 @@ std::shared_ptr<PublisherData> PublisherData::make(
   }
 
   using AdvancedPublisherOptions = zenoh::ext::SessionExt::AdvancedPublisherOptions;
+  using SampleMissDetectionOptions = AdvancedPublisherOptions::SampleMissDetectionOptions;
   auto adv_pub_opts = AdvancedPublisherOptions::create_default();
 
   if (adapted_qos_profile.durability == RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL) {
@@ -126,8 +127,9 @@ std::shared_ptr<PublisherData> PublisherData::make(
       // If RELIABLE + TRANSIENT_LOCAL activate sample miss detection for subscriber
       // to detect missed samples and retrieve those from the Publisher cache.
       // HeartbeatSporadic is used to prevent excessive background traffic
-      adv_pub_opts.sample_miss_detection.emplace().heartbeat =
-        AdvancedPublisherOptions::SampleMissDetectionOptions::HeartbeatSporadic{
+      adv_pub_opts.sample_miss_detection = SampleMissDetectionOptions{};
+      adv_pub_opts.sample_miss_detection->heartbeat =
+        SampleMissDetectionOptions::HeartbeatSporadic{
         SAMPLE_MISS_DETECTION_HEARTBEAT_PERIOD};
     }
   }

--- a/rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp
+++ b/rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp
@@ -175,6 +175,7 @@ bool SubscriptionData::init()
   sess_ = context_impl->session();
 
   using AdvancedSubscriberOptions = zenoh::ext::SessionExt::AdvancedSubscriberOptions;
+  using RecoveryOptions = AdvancedSubscriberOptions::RecoveryOptions;
   auto adv_sub_opts = AdvancedSubscriberOptions::create_default();
 
   // By default, this subscription will receive publications from publishers within and outside of
@@ -200,8 +201,8 @@ bool SubscriptionData::init()
       // Activate recovery of lost samples.
       // This requires the Publisher to have sample_miss_detection configured,
       // which is the case for a RELIABLE + TRANSIENT_LOCAL Publisher.
-      adv_sub_opts.recovery.emplace().last_sample_miss_detection =
-        AdvancedSubscriberOptions::RecoveryOptions::Heartbeat{};
+      adv_sub_opts.recovery = AdvancedSubscriberOptions::RecoveryOptions{};
+      adv_sub_opts.recovery->last_sample_miss_detection = RecoveryOptions::Heartbeat{};
     }
   }
 

